### ISSUE TITLE
Newtype declarations conflict with using "type" as a parameter name

### ISSUE THREAD

**Issue by SantiagoBautista** [p4lang/p4c#1335]
The psa.p4 file uses the lexeme `type` as a parameter name (at lines 404, 433, 465 and 490, in the constructors of externs Counter, DirectCounter, Meter, DirectMeter respectively).

But since newtype declarations are now possible, the lexer will probably interpret the lexeme `type` as the `TYPE` token, and not as an IDENTIFIER called "type". This should result in the lexeme `type` not being recognizable as a name, and hence not being recognizable as the name of a parameter, like it is used in psa.p4 .

I see two solutions to this:
+ Either we say that `type` can no longer be used as a parameter name as it is now a keyword of the language, and we hence change psa.p4
+ Or we change the parser and the grammar so that the TYPE token can also be recognized as a name (as are APPLY, KEY, ACTIONS and STATE tokens).

This second solution needs care: I tried implementing it in an interpreter I am coding, and I found out that if TYPE token is allowed to be a nonTypeName, this raises a conflict in the parser generator between function declarations and newtype declarations. (I can explain what this conflict is if someone is interested). Because of that, I had to extend the `name` rule instead of extending the `nonTypeName` rule as it is done for tokens APPLY, KEY, ACTIONS and STATE.

There is something that I don't understand about this, and I would appreciate help from people developing the compiler. I expect the above problem to generate a parsing error, but the compiler seems to pass the tests that use psa.p4. I tried to test the compiler on a single file instead of running the test harness, and I couldn't, because the psa backend in not yet finished.
So my question is: What exactly do the tests do for the files using psa.p4, since the psa backend of the compiler is not yet finished?

**Comment by mbudiu-vmw** [comment]
Yes, the "name" non-terminal can include TYPE (and a few other keywords).
This allows keywords to be used as an identifier in selected places where it is not ambiguous.
So the grammar is actually good.

**Comment by mbudiu-vmw** [comment]
If you run a compiler with `--validate` it will run a file just through the front-end; this should accept all syntactically correct programs for all possible targets.
You can run the compiler with `-v` to see the list of passes that get executed.

**Comment by SantiagoBautista** [comment]
The last time I had looked at the compiler I hadn't seen TYPE in the keywords that can be names, I guess I had forgotten to pull the latest version. My bad.

But this is not yet in the grammar (I just checked) . I will open a pr that updates the grammar.

Thank you for answering my question !

This gives another question, that is why you didn't encounter the same conflict that I encountered when allowing TYPE to be a nonTypeName. I will open another issue detailing what the conflict is and giving an example.
